janet-http/http-service/src/main/java/io/techery/janet/HttpActionService.java

```diff
@@ -85,48 +87,79 @@ public HttpActionService(String baseUrl, HttpClient client, Converter converter)
     @Override protected <A> void sendInternal(ActionHolder<A> holder) throws HttpServiceException {
         callback.onStart(holder);
         A action = holder.action();
-        runningActions.add(action);
+        putRunningAction(action);
         final ActionHelper<A> helper = getActionHelper(action.getClass());
         if (helper == null) {
             throw new JanetInternalException("Something was happened with code generator. Check dependence of janet-http-compiler");
         }
         RequestBuilder builder = new RequestBuilder(baseUrl, converter);
         Response response;
+        Request request = null;
         try {
             builder = helper.fillRequest(builder, action);
-            Request request = builder.build();
-            throwIfCanceled(action);
-            response = client.execute(request, new ActionRequestCallback<A>(holder) {
-                private int lastProgress;
-
-                @Override public void onProgress(int progress) {
-                    if (progress > lastProgress + PROGRESS_THRESHOLD) {
-                        callback.onProgress(holder, progress);
-                        lastProgress = progress;
+            request = builder.build();
+            putRunningRequest(action, request);
```

Comment:
not sure why we putting request with action again after we've already put action (see `putRunningRequest` calls `putRunningAction`)